### PR TITLE
[WIP] Introduce max-age client-side connection option

### PR DESCRIPTION
**Pull request by julnamoo**
Motivation
---
 When client opens keepalive connections, a client sends requests
 to the first server only unless the connection is closed by idle timeout.
 This behavior leads the reqeust imbalance among servers.
 In order to distribute requests evenly, a client could refresh its connection
 regularly to reconnet to other server actively.

Modifications
---
 - Introduce `maxAge` clientOption to `ClientFactoryBuilder`,
 `ClientFactoryOptions` and `ClientFactoryOption`
 - Add a jvm flag option to override the option

- [ ] Add test

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/client/ClientFactoryBuilder.java [hunk 1]
```diff
@@ -427,6 +427,16 @@ public ClientFactoryBuilder idleTimeoutMillis(long idleTimeoutMillis) {
         return this;
     }
 
+    /**
+     * Sets the max age of a socket connection in milliseconds. The connection is closed
+     * after this amount of time of creation.
+     */
+    public ClientFactoryBuilder maxAgeMillis(long maxAgeMillis) {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/ClientFactoryBuilder.java:434]: - `maxConnectionAgeMillis()`?
- Could you also add `maxConnectionAge(Duration)`?

File: core/src/main/java/com/linecorp/armeria/client/ClientFactoryOption.java [hunk 2]
```diff
@@ -179,6 +179,12 @@
     public static final ClientFactoryOption<Long> IDLE_TIMEOUT_MILLIS =
             define("IDLE_TIMEOUT_MILLIS", Flags.defaultClientIdleTimeoutMillis());
 
+    /**
+     * The max-age of a socket connection in milliseconds.
+     */
+    public static final ClientFactoryOption<Long> MAX_AGE_MILLIS =
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/ClientFactoryOption.java:185]: - `MAX_CONNECTION_AGE_MILLIS`
- Could you add: `The connection is closed after this amount of time of creation.`

File: core/src/main/java/com/linecorp/armeria/client/ClientFactoryOptions.java [hunk 3]
```diff
@@ -218,6 +218,10 @@ public long idleTimeoutMillis() {
         return get(ClientFactoryOption.IDLE_TIMEOUT_MILLIS);
     }
 
+    public long maxAgeMillis() {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/ClientFactoryOptions.java:221]: Missing Javadoc

File: core/src/main/java/com/linecorp/armeria/server/Http1ServerKeepAliveHandler.java [hunk 4]
```diff
@@ -42,4 +42,9 @@ protected boolean hasRequestsInProgress(ChannelHandlerContext ctx) {
         final HttpServer server = HttpServer.get(ctx);
         return server != null && server.unfinishedRequests() != 0;
     }
+
+    @Override
+    protected void closeMaxAgedConnection(ChannelHandlerContext ctx) {
+        // silently doing nothing
```
Comment (julnamoo) [inline core/src/main/java/com/linecorp/armeria/server/Http1ServerKeepAliveHandler.java:48]: Left as doing nothing to wait for the further implementation

**Comment by trustin** [comment]
@julnamoo Gentle ping 😉

**Comment by julnamoo** [comment]
Discussed to deal with by #3267 .
Let me close this PR, thanks for working, @ikhoon !

**Pull request closed without merging**